src/models/CDD.java

```diff
@@ -260,106 +304,125 @@ private Guard toBoolGuards()
     }
 
     public long getPointer() {
-        return pointer;
+        // We have to use getValue as we don't want to return null if the pointer property is dirty.
+        return pointerProperty.getValue();
+    }
+
+    private void setPointer(long newPointer) {
+        pointerProperty.markAsDirty();
+        pointerProperty.set(newPointer);
     }
 
     public int getNodeCount()
             throws NullPointerException {
         checkForNull();
-        return CDDLib.cddNodeCount(pointer);
+        return CDDLib.cddNodeCount(getPointer());
     }
 
     public CDDNode getRoot()
             throws NullPointerException {
         checkForNull();
-        long nodePointer = CDDLib.getRootNode(this.pointer);
+        long nodePointer = CDDLib.getRootNode(getPointer());
         return new CDDNode(nodePointer);
     }
 
     public CddExtractionResult extract()
             throws NullPointerException, CddNotRunningException {
-        if (isExtractionDirty) {
+        return extractionProperty.trySet(() -> {
             checkIfNotRunning();
             checkForNull();
-            extraction = new CddExtractionResult(
-                    CDDLib.extractBddAndDbm(pointer)
+            return new CddExtractionResult(
+                    CDDLib.extractBddAndDbm(reduce().getPointer())
             );
-            isExtractionDirty = false;
-        }
-
-        return extraction;
+        });
     }
 
     public boolean isBDD()
             throws NullPointerException {
-        checkForNull();
-        if (isBddDirty) {
-            // CDDLib.isBDD does not recognise cddFalse and cddTrue as BDDs
-            isBdd = isFalse() || isTrue() || CDDLib.isBDD(this.pointer);
-        }
-
-        return isBdd;
+        return isBddProperty.trySet(() -> {
+            checkForNull();
+            checkIfNotRunning();
+            // TODO: Make this terminal checks.
```

Comment:
This `TODO` is intentionally here? Is there a related issue so we don't forget to do this?